drivers/clock_control/nrf5_power_clock.c

```diff
@@ -141,15 +143,35 @@ static int _k32src_start(struct device *dev, clock_control_subsys_t sub_system)
 {
 	u32_t lf_clk_src;
 	u32_t intenset;
+	u32_t imask;
 
-	/* TODO: implement the ref count and re-entrancy guard, if a use-case
-	 * needs it.
+	/* If the LF clock is already started, but wasn't initialized with
+	 * this function, allow it to run once. This is needed because if a
+	 * soft reset is triggered while watchdog is active, the LF clock will
+	 * already be running, but won't be configured yet (watchdog forces LF
+	 * clock to be running).
+	 *
+	 * Afterwards, allow the function to run at any time the LF clock is
+	 * stopped.
 	 */
+	imask = irq_lock();
 
-	if ((NRF_CLOCK->LFCLKSTAT & CLOCK_LFCLKSTAT_STATE_Msk)) {
+	if (k32src_guard) {
+		irq_unlock(imask);
+		goto lf_already_started;
+	}
+
+	if (k32src_initialized_once &&
```

Comment:
if already initialized once then irq_unlock  and goto lf_already_started. Refer to line 35 to be consistent. Check for  h/w stat is not needed if the design in _m16src_start is followed.

Reply:
I was doing that at first for consistency, but I changed it later for a little more correctness. Doesn't that add a second `if` (equal to the `goto` check) which, if pre-empted away with a _just right_ timing, may return a different value, returning `-EINPROGRESS` instead of 0? That may be too minor an issue to worry about though, even if I would prefer it to be bulletproof...

About the h/w stat check, I think it's needed here. The design used in `_m16src_start` doesn't apply here, because we have no `stop` function implemented. Because of this, there's no need for a ref count -- it would never decrease. If we want to disallow callers to call it more than once, that's fine. My intention was to allow the caller to call the function at least once (even if the peripheral is already on, to configure it) and, after this, allow the caller to call it every time the peripheral was disabled. The use case for this is if the user manually stops it outside of the driver for some reason. What do you think?

Reply:
_m16src_start enters with irq_lock, hence I dont see what gets pre-empted away. If _m16src_start is not bulletproof it needs fixing.

(k32src_initialized_once == 1) implies (NRF_CLOCK->LFCLKSTAT & CLOCK_LFCLKSTAT_STATE_Msk) != 0. Else something is wrong in following code that the code has not started the LF clock after the first time k32src_initialized_once  was set. Do you agree?

Reply:
@cvinayak, `_m16src_start` is fine. From what I can see, the ref count handles all cases correctly.

About the 32 KHz, looking from the point of view of the driver, your assessment is correct and I agree. What I meant is that, if the user, for some reason, wanted to control the peripheral directly (i.e `NRF_CLOCK->TASKS_LFCLKSTOP = 1`), the driver wouldn't work as they expected anymore. Example:

```
struct device *clock;

clock = device_get_binding(CONFIG_CLOCK_CONTROL_NRF5_K32SRC_DRV_NAME);
if (!clock) {
    return -1;
}

clock_control_on(clock, (void *) CLOCK_CONTROL_NRF5_K32SRC);

NRF_CLOCK->TASKS_LFCLKSTOP = 1;

/* wait until clock stops */

/* wouldn't work as expected by the user; the 32 KHz clock source wouldn't 
 * turn on/be configured again
 */
clock_control_on(clock, (void *)CLOCK_CONTROL_NRF5_K32SRC);
```

I do concede that this is a very niche use case, but I like when things don't fail in unexpected conditions. However, seeing as the high frequency clock source doesn't support this, there seems to be no need to support it in the low frequency clock source. Consistency is king.

Then, do you agree with the following course of action?

1) Remove the guard, as it is not needed anymore. The initialized once check already handles this case well;
2) Remove the h/w stat check, keeping only the k32src_initialized_once check. If it fails, goto already started.

An alternative course of action would be to implement the stop function and the ref count. If you think that is better, I would gladly do it. My initial focus was only on solving the bug, but I could expand this PR to include the stop feature.

Thanks for the attention,
Thiago

Reply:
Hi Thiago,

I agree to your 2 step resolution. This is the simplest.

Lets not add ref count and stop implementations, as it will not be used yet (as it was before). I agree we solve the bug.

Thanks and regards,
Vinayak

Reply:
Done!